File: 1_Review/javascript/8_generator_expressions.js
// 8) Generators (JS has generator functions)
// Run: node 8_generator_expressions.js

// Generator function using `yield`
function* generatorFunction() {
  yield 'START';
  yield 'PROCESSING';
  yield 'DONE';
}

let genFunc = generatorFunction();

console.log('Print using generator function and next()');
console.log(genFunc.next().value);
console.log(genFunc.next().value);
console.log(genFunc.next().value);
console.log(genFunc.next().value ?? 'No more states');

console.log('\nPrint using generator function in for..of loop');
genFunc = generatorFunction();
for (const state of genFunc) {
  console.log(state);
}

// List of numbers 0 through 9
const numbers = Array.from({ length: 10 }, (_, i) => i);

// Generator "expression" style: create a generator that yields squares
function* squares(items) {
  for (const x of items) {
    yield x ** 2;
  }
}

const genExp = squares(numbers);

console.log('\nPrint using generator and next()');
console.log(genExp.next().value);
console.log(genExp.next().value);
console.log(genExp.next().value);

console.log('\nPrint remaining squares using for..of loop');
for (const item of genExp) {
  console.log(item);
}

File: 1_Review/python/8_generator_expressions.py


# Generator function using yield
def generator_function():
    yield "START"
    yield "PROCESSING"
    yield "DONE"

gen_func = generator_function()

print('Print using generator function and next()')
# next prints the next item returned by the generator
print(next(gen_func))
print(next(gen_func))
print(next(gen_func))
#print(next(gen_func))
print(next(gen_func,'No more states'))

gen_func = generator_function()

for state in gen_func:
    print(state)

for state in generator_function():
    print(state)


# print('Print using generator function in for loop')
# # Using a loop to get remaining values
# for item in gen_func:
#     print(item)

# List of numbers 0 through 9
numbers = [x for x in range(10)]

# Generator expression
gen_exp = (x**2 for x in numbers)

print('Print using generator expression and next()')
# next prints the next item returned by the generator expression
print(next(gen_exp))
print(next(gen_exp))
print(next(gen_exp))
print(next(gen_exp))
print(next(gen_exp))
print(next(gen_exp))
print(next(gen_exp))
print(next(gen_exp))
print(next(gen_exp))
print(next(gen_exp))
# #print(next(gen_exp))
# #print(next(gen_exp,'No more squares'))

print('Print using generator expression in for loop')
# # Using a loop to get remaining values
for item in gen_exp:
    print(item)